dagshub/data_engine/model/datasource.py

```diff
@@ -158,18 +175,26 @@ def global_as_of_timestamp(self) -> Optional[int]:
     def source(self) -> "DatasourceState":
         return self._source
 
-    def clear_query(self):
+    def clear_query(self, reset_to_dataset=True):
         """
-        Clear the attached query. Next time this datasource gets queried, you'll get all of the datapoints.
+        Clear the attached query.
 
-        .. note::
-            This function always clears the query.
-            That means that if you use it on a dataset, the whole query will get cleared, and not just your changes.
+        Args:
+            reset_to_dataset: If ``True`` and this Datasource was saved as a dataset, reset to the query in the dataset,
+                instead of clearing the query completely.
         """
-        self._query = DatasourceQuery()
+        if reset_to_dataset and self.assigned_dataset is not None and self.assigned_dataset.query is not None:
+            self._query = self.assigned_dataset.query.__deepcopy__()
+        else:
+            self._query = DatasourceQuery()
 
     def __deepcopy__(self, memodict={}) -> "Datasource":
         res = Datasource(self._source, self._query.__deepcopy__(), self._select, self._global_as_of)
+        res.assigned_dataset = self.assigned_dataset
+
+        # Carry over the update context, that way we'll keep track of the metadata being uploaded
+        res._implicit_update_ctx = self._implicit_update_ctx
```

Comment:
What about the explicit context?

Reply:
that one doesn't really matter because it only gets uploaded once you exit the context.
We're not losing the reference to it, so the explicit context is persisting.

So examples for code where the implicit context passing is relevant

Implicit context passing (doesn't work currently):
```python
dps = ds.all()
dp = dps[0]
dp["field1"] = 2

queried = ds["sub_query"] == "fsadfasfd"
dp2 = queried.all()[0]
dp2["field2"] = 3

dp2.save()  # Will save both the dp and the dp2 change
```

Explicit context passing:
```python
with ds.update_metadata() as ctx:
  ctx.update_metadata(...)
  queried = ds[..] == ...
  with queried.update_metadata() as ctx2:
    ctx2.update_metadata(...)
  # Uploads ctx2
#Uploads ctx
...
```